marconi-core/src/Marconi/Core/Experiment/Transformer/WithCatchup.hs

```diff
@@ -0,0 +1,177 @@
+  -> Word
+  -- ^ Maximal number of elements to send to the indexer in one batch
+  -> indexer event
+  -- ^ the underlying indexer
+  -> WithCatchup indexer event
+withCatchup computeDistance bypassDistance batchSize =
+  WithCatchup . IndexWrapper (CatchupConfig computeDistance bypassDistance batchSize 0 [])
+
+deriving via
+  (IndexWrapper CatchupConfig indexer)
+  instance
+    (IsSync m event indexer) => IsSync m event (WithCatchup indexer)
+
+deriving via
+  (IndexWrapper CatchupConfig indexer)
+  instance
+    (Closeable m indexer) => Closeable m (WithCatchup indexer)
+
+deriving via
+  (IndexWrapper CatchupConfig indexer)
+  instance
+    (Queryable m event query indexer) => Queryable m event query (WithCatchup indexer)
+
+caughtUpIndexer :: Lens.Lens' (WithCatchup indexer event) (indexer event)
+caughtUpIndexer = catchupWrapper . wrappedIndexer
+
+instance IndexerTrans WithCatchup where
+  type Config WithCatchup = CatchupConfig
+
+  wrap cfg = WithCatchup . IndexWrapper cfg
+
+  unwrap = caughtUpIndexer
+
+{- | A typeclass that allows an indexer with a @WitchCatchup@ transformer to configure
+ the behaviour of this transformer
+-}
+class HasCatchupConfig indexer where
+  catchupBypassDistance :: Lens.Lens' (indexer event) Word
+  catchupBatchSize :: Lens.Lens' (indexer event) Word
+
+instance {-# OVERLAPPING #-} HasCatchupConfig (WithCatchup indexer) where
+  catchupBypassDistance = catchupWrapper . wrapperConfig . configCatchupBypassDistance
+  catchupBatchSize = catchupWrapper . wrapperConfig . configCatchupBatchSize
+
+instance
+  {-# OVERLAPPABLE #-}
+  (IndexerTrans t, HasCatchupConfig indexer)
+  => HasCatchupConfig (t indexer)
+  where
+  catchupBypassDistance = unwrap . catchupBypassDistance
+  catchupBatchSize = unwrap . catchupBatchSize
+
+instance
+  {-# OVERLAPPABLE #-}
+  (IndexerMapTrans t, HasCatchupConfig indexer)
+  => HasCatchupConfig (t indexer output)
+  where
+  catchupBypassDistance = unwrapMap . catchupBypassDistance
+  catchupBatchSize = unwrapMap . catchupBatchSize
+
+catchupDistance :: Lens.Lens' (WithCatchup indexer event) (Point event -> event -> Word)
+catchupDistance = catchupWrapper . wrapperConfig . configDistanceComputation
+
+catchupBuffer :: Lens.Lens' (WithCatchup indexer event) [Timed (Point event) (Maybe event)]
+catchupBuffer = catchupWrapper . wrapperConfig . configCatchupBuffer
+
+catchupBufferLength :: Lens.Lens' (WithCatchup indexer event) Word
+catchupBufferLength = catchupWrapper . wrapperConfig . configCatchupBufferLength
+
+resetBuffer :: WithCatchup indexer event -> WithCatchup indexer event
+resetBuffer = (catchupBufferLength .~ 0) . (catchupBuffer .~ [])
+
+instance (IsIndex m event indexer) => IsIndex m event (WithCatchup indexer) where
+  index timedEvent indexer =
+    let bufferIsFull ix = (ix ^. catchupBufferLength) >= (ix ^. catchupBatchSize)
+        pushEvent ix =
+          ix
+            & catchupBuffer %~ (timedEvent :)
+            & catchupBufferLength +~ 1
```

Comment:
You keep the length in memory for efficiency reasons?

Reply:
yes, it isn't a huge perf penalty but I feel dirty when I go through several thousands of elements just to get a length.